### ISSUE TITLE
[RFC] DDP Communication Hook

### ISSUE THREAD

**Issue by pritamdamania87** [pytorch/pytorch#39272]
# Motivation

There are several algorithms like GossipGrad and gradient compression which involve different communication strategies for parameter syncs while running Distributed DataParallel training. The high level idea here is to provide a flexible hook to users where they can specify how DDP aggregates gradients across multiple workers. Apart from allowing experimentation with ideas mentioned above, this hook would be very useful for researchers to try out new ideas.

# Proposal

Provide a hook that does all the necessary processing for communication and returns a Future indicating the completion of the task

```
// DDP APIs
ddp.register_comm_hook(
    state: object, 
    hook: callable)

// DDP combines multiple parameters into a bucket before doing an allreduce.
class GradBucket:
  Tensor bucket_data
  // Optionally in the future this can be enhanced with parameter to bucket 
  // mappings as well.
  

// The state object is passed to the hook and can be used 
// to maintain and update any state information that users would like to 
// maintain as part of the training process. Examples: error feedback in 
// gradient compression, peers to communicate with next in GossipGrad etc. 

// Hook signature
// The hook is passed the state object and the DDP grad bucket. 
// This function is called once the bucket is ready. The hook can perform
// whatever processing is needed and return a Future indicating 
// completion of any async work (ex: allreduce). If the hook doesn't 
// perform any communication, it can also just return a completed Future.
// The Future should hold the new value of the bucket.
// Once a bucket is ready, c10d reducer would call this hook and use the 
// bucket returned by the Future as the new bucket and would use this 
// bucket to copy grads to individual parameters.
def ddp_comm_hook(state: object, bucket: GradBucket): -> Future
```
# Examples

## Gradient Compression
```
ddp.register_comm_hook(
    state=None, 
    hook: fp16_compress)
    
def fp16_compress(state: object, bucket: Tensor): -> Future
    compressed_bucket = bucket.to(torch.float16)
    state.compressed_buckets[bucket] = compressed_bucket
    work = dist.allreduce(compressed_bucket)
    # NOTE: We also provide an API called "get_future" to retrieve a future
    associated with the completion of c10d.ProcessGroupNCCL.work.
    allreduce_future =work.get_future()
    
    def decompress(fut):
        compressed_bucket = fut.wait()
        return compressed_bucket.to(torch.float32)
        
    return allreduce_future.then(decompress) 
```

## GossipGrad

```
class GossipGradState:
    def get_next_peers() -> List[Peer]
    # Modifies passed in grad in-place after gossiping. Returns a future 
    # indicating completion of the work.
    def gossip_grads(peers: List[Peer], grad: Tensor) -> Future
    
ddp.register_comm_hook(
    state=GossipGradState(...), 
    hook: gossip_grad_hook)
    
def gossip_grad_hook(state: object, bucket: Tensor): -> Future
    peers = state.get_next_peers()
    return gossip_grads(peers, bucket)
```

## Sub-tasks:
- [x] Create a `Future` type `comm_hook` in the `reducer` and register the DDP communication hook in the Python API. We avoid using `py::object` directly in `reducer.cpp`, so `PythonCommHook` is an instance of abstract class `CommHookInterface` and define the hook as `std::unique_ptr<CommHookInterface> comm_hook_`. **See #40848 **
- [x] Reducer calls and waits for `future_work` instead of `work` if the hook was registered. Then it obtains result of the `future_work`. **See #40848 **
- [ ] Provide `get_future` API to retrieve a `Future` associated with the completion of `c10d.ProcessGroupNCCL.work`. **See #42335 **

Warning: `get_future` API is currently only supported by NCCL backend. Please refer #42048 for gloo support.

cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @xush6528 @osalpekar

**Comment by vineeths96** [comment]
Any updates?

**Comment by sinannasir** [comment]
> Any updates?

Hi @vineeths96 , we expect this to be functional by the end of this week or sometime next week. It is now prototyped in PR #40848 if you wanna take a look. The next goal is to finish an API that converts ProcessGroup::Work to utils::Future. Then, DDP Comm hook should be ready to try some compression techniques out.

**Comment by vineeths96** [comment]
Hi @sinannasir 

Thank you for your reply.

That's some really good work out there. I experiment with different gradient compression techniques. I am eagerly waiting to try out some compression algorithms using this new development.

**Comment by vineeths96** [comment]
@sinannasir I went through the latest commit. However, I need some clarification regarding the implementation.  

In the [_register_comm_hook](https://github.com/pytorch/pytorch/blob/65bd38127a34d428915c88507878b4735edf005f/torch/nn/parallel/distributed.py#L606) function, is the hook function called after the gradients are accumulated into the buckets? If yes, then would not compression impact only in the allreduce process and not in accumulation into the buckets process?

More specifically, compression schemes such as [QSGD](https://arxiv.org/abs/1610.02132) and [TernGrad](https://arxiv.org/abs/1705.07878) which perform operations on gradients before communication or accumulation into buckets cannot be implemented?

**Comment by sinannasir** [comment]
After PRs #40848 and #41596, we enabled the communication hook where user can specify how DDP aggregates gradients across multiple workers. Note that `get_future` API is currently only supported by NCCL backend. Please refer #42048 for Gloo support.

**Comment by pritamdamania87** [comment]
Re-opening since we had to revert the original PR. @sinannasir feel free to close this when we land the second version.

**Comment by sinannasir** [comment]
Closing this issue again. We re-implemented get_future API using `FutureNCCL` #42335.

Some DDP Communication hook examples were also landed in #43310

**Comment by david-macleod** [comment]
Is this feature targeting torch 1.7?

**Comment by pritamdamania87** [comment]
@david-macleod Yes, this feature would be in torch 1.7, but it will be marked as "prototype" to begin with since it currently only supports the NCCL backend and single process single device mode of DDP. Would love to hear your feedback on any improvements we can make to this hook for use cases you have in mind.

**Comment by david-macleod** [comment]
Thanks for your reply, that's great news, do you know if it will be in the nightly builds soon? The use case we have is actually covered exactly by one of the examples you have provided (f16 compression).

**Comment by pritamdamania87** [comment]
@david-macleod It should already be in the nightly build, you can find the API to register a hook here: https://github.com/pytorch/pytorch/blob/69839ea3f6680d068db38795afe03ae4b8a30962/torch/nn/parallel/distributed.py#L863-L943 and there are some examples here: https://github.com/pytorch/pytorch/tree/master/torch/distributed/algorithms/ddp_comm_hooks.

**Comment by david-macleod** [comment]
@pritamdamania87 I am starting to use the `fp16_compress_hook` and noticing that an extra copy of the gradient tensors are kept in memory. In the example below the size of the parameters are 8MB so the expected allocated memory is 24MB (parameter tensors + gradient tensors + DDP buffers), and I have confirmed this is the case in the standard setting. 

However when using the hook this increased to 32MB, presumably caused by the `bucket.get_tensors()[0].to(torch.float16)` copy, but I would have expected this memory to be released after the synchronization was complete, but that does not seem to be the case. Do you know of any way to avoid this extra copy as it becomes an issue when dealing with very large models.

Many thanks

```py
import torch
from torch.nn.parallel import DistributedDataParallel as DDP

device = "cuda"
model = torch.nn.Linear(512, 4096).to(device)

model = DDP(model, device_ids=[torch.cuda.current_device()])
model._register_comm_hook(state=None, hook=fp16_compress_hook) 
optimizer = torch.optim.SGD(model.parameters(), lr=1e-4)

for _ in range(5):
    y = model(torch.randn(64, 512, device=device)).mean()
    y.backward()
    optimizer.step()
    optimizer.zero_grad()

print(torch.cuda.memory_allocated() / (1024 ** 2))
```

**Comment by david-macleod** [comment]
Have opened an issue here https://github.com/pytorch/pytorch/issues/45968

**Comment by SciPioneer** [comment]
Updates:

1. DDP comm hooks are formally available since PyTorch 1.8: https://pytorch.org/docs/master/ddp_comm_hooks.html
2. DDP comm hooks are also upstreamed to PyTorch Lightning: https://github.com/PyTorchLightning/pytorch-lightning/pull/6736
